s3270/x3270if.man
'\" t
.TH X3270IF 1 "17 February 2020"
.SH "NAME"
x3270if \- command interface to
x3270, c3270 and s3270
.SH "SYNOPSIS"
\fBx3270if\fP [option]... [ \fIaction\fP ]
.br
\fBx3270if \-i\fP
.SH "DESCRIPTION"
\fBx3270if\fP provides an interface between scripts and
the 3270 emulators
\fIx3270\fP, \fIc3270\fP and \fIs3270\fP.
.LP
\fBx3270if\fP operates in one of two modes.
In \fBaction mode\fP, it passes a single action and parameters to the
emulator for execution.
The result of the action is written to standard output, along with the
(optional) status of the emulator.
(The action is optional as well, so that \fBx3270if\fP can just reports the
emulator status.)
In \fBiterative mode\fP, it forms a continuous conduit between a script and
the emulator.
.LP
The \fIaction\fP takes the form:
.IP
\fIaction-name\fP(\fIparam\fP[,\fIparam\fP]...)
.LP
The parentheses are manatory, and usually must be quoted when \fBx3270if\fP is
called from a shell script.
.LP
If any \fIparam\fP contains a space or comma, it must be surrounded by
double quotes.
.SH "OPTIONS"
.TP
\fB\-s\fP \fIfield\fP
Causes \fBx3270if\fP to write to stdout the value of one of the
emulator status fields.
\fIField\fP is an integer in the range 0 through 11.
The value 0 is a no-op and is used only to return exit status indicating the
state of the emulator.
The indices 1-11 and meanings of each field are documented on the
\fIx3270-script\fP(1) manual page.
If an \fIaction\fP is specified as well, the status field is written after the
output of the action, separated by a newline.
The \fB\-s\fP option is mutually exclusive with the
\fB\-S\fP and \fB\-i\fP options.
.TP
\fB\-S\fP
Causes \fBx3270if\fP to write to stdout the value of all of the
emulator status fields.
If an \fIaction\fP is specified as well, the status fields are written after the
output of the action, separated by a newline.
The \fB\-S\fP option is mutually exclusive with the
\fB\-s\fP and \fB\-i\fP options.
.TP
\fB\-i\fP
Puts \fBx3270if\fP in iterative mode.
Data from \fBx3270if\fP's standard input is copied to the
emulator's script input; data from the
emulator's script output is copied to
\fBx3270if\fP's standard output.
The \fB\-i\fP option is mutually exclusive with the
\fB\-s\fP and \fB\-S\fP options.
\fBx3270if\fP
runs until it detects end-of-file on its standard input or on the
output from the emulator.
(This mode exists primarily to give \fIexpect\fP(1)
a process to run, on systems which do not support bidirectional pipes.)
.TP
\fB\-p\fP \fIprocess-id\fP
Causes \fIx3270if\fP to use a Unix-domain socket to connect to the emulator,
rather than pipe file descriptors given in environment variables.
The emulator must have been started with the \fB\-socket\fP option.
.TP
\fB\-t\fP \fIport\fP
Causes \fIx3270if\fP to use a TCP socket to connect to the emulator,
rather than pipe file descriptors given in environment variables.
The emulator must have been started with the \fB\-scriptport\fP
option.
.TP
\fB\-v\fP
Turns on verbose debug messages, showing on stderr the literal data that is
passed between the emulator and \fBx3270if\fP.
.SH "EXIT STATUS"
In action mode, if the requested \fIaction\fP succeeds,
\fBx3270if\fP exits with status 0.
If the action fails, \fBx3270if\fP exits with status 1.
In iterative mode, \fBx3270if\fP
exits with status 0 when it encounters end-of-file.
If there is an operational error within \fBx3270if\fP
itself, such as a command-line syntax error, missing environment
variable, or an unexpectedly closed pipe,
\fBx3270if\fP exits with status 2.
.SH "ENVIRONMENT"
When a script is run as a child process of one of the emulators via the
\fBScript\fP action, the emulator passes information about how to control it
in environment variables.
.LP
On Unix, the emulator process creates a pair of pipes for communication with
the child script process.
The values of the file descriptors for these pipes are encoded as text
in two environment variables:
.TP
\fBX3270OUTPUT\fP
Output from the emulator, input to the child process.
.TP
\fBX3270INPUT\fP
Input to the emulator, output from the child process.
.LP
When an emulator is started with the
\fB\-scriptport\fP option, the 
emulator will pass the scriptport port number
encoded as text in the \fBX3270PORT\fP environment variable.
\fIx3270if\fP will use that value as if it had been passed to it via the
\fB\-t\fP option.
\fBX3270PORT\fP takes precedence over
\fBX3270OUTPUT\fP and \fBX3270INPUT\fP.
.SH "SEE ALSO"
x3270(1), c3270(1),
s3270(1),
x3270-script(1)
.SH "COPYRIGHT"
Copyright 1999-2009, 2020 Paul Mattes.
.br
All rights reserved.
.LP
Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

.TP
*
Redistributions of source code must retain the above copyright
notice, this list of conditions and the following disclaimer.
.TP
*
Redistributions in binary form must reproduce the above copyright
notice, this list of conditions and the following disclaimer in the
documentation and/or other materials provided with the distribution.
.TP
*
Neither the names of Paul Mattes nor the names of his contributors
may be used to endorse or promote products derived from this software
without specific prior written permission.

.LP
THIS SOFTWARE IS PROVIDED BY PAUL MATTES `AS IS' AND ANY EXPRESS OR IMPLIED
WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
EVENT SHALL PAUL MATTES BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


Releases
3.3.10alpha1	575
3.3.10alpha2	589
3.3.10ga3	593
3.3.10ga4	609
3.3.10ga5	673
3.3.11beta2	684
3.3.11beta3	687
3.3.11beta4	697
3.3.11ga6	711
3.3.12alpha3	732
3.3.12ga7	760
3.3.12b9	772
3.3.12ga10	782
3.3.12ga11	788
3.3.12ga12	829
3.3.12ga13	888
3.3.13alpha1	954
3.3.13alpha2	973
3.3.13beta3	976
3.3.13beta4	1023
3.3.13beta5	1065
3.3.13ga6	1080
3.3.13ga7	1084
3.3.14alpha1	1133
3.3.14alpha2	1142
3.3.14alpha3	1151
3.3.14alpha4	1153
3.3.14alpha5	1179
3.3.14ga6	1204
3.3.14ga7	1240
3.3.14beta8	1248
3.3.14ga9	1258
3.3.14ga11	1271
3.3.15alpha1	1316
3.3.15beta2	1342
3.3.15ga4	1373
3.3.15ga5	1378
3.3.15ga6	1400
3.3.15ga7	1432
3.3.15ga8	1469
3.4alpha1	1825
3.4alpha2	1867
3.3.15ga9	1878
3.4beta3	1880
3.4ga4		1897
3.4ga5		1931
3.5alpha1	1939
3.4ga6		1958
3.4ga7		1979
3.4ga8		1998


Common/x3270if.py
#!/usr/bin/env python3
# Simple Python version of x3270if

"""Python interface to x3270 emulators"""

import io
import os
import socket
import sys
import subprocess
import time

_backslashChars = '\\"'
_quoteChars = _backslashChars + ' ,()'

def Quote(arg):
    """Quote an argument in action syntax

       Args:
          arg (str): Argument to format
       Returns:
          str: Formatted argument
    """
    # Backslashes and double quotes need backslashes in front of them.
    # Other syntax markers (space, comma, paren) just need the argument in
    #  double quotes.
    if (not any(ch in arg for ch in _quoteChars)): return arg
    return '"' + ''.join('\\' + ch if ch in _backslashChars else ch for ch in arg) + '"'

class ActionFailException(Exception):
    """x3270if action failure"""
    def __init__(self,msg):
        Exception.__init__(self,msg)

class StartupException(RuntimeError):
    """x3270if unable to start s3270"""
    def __init__(self,msg):
        RuntimeError.__init__(self,msg)

class _x3270if():
    """Abstract x3270if base class"""
    def __init__(self,debug=False):
        """Initialize an instance

           Args:
              debug (bool): True to trace debug info to stderr.
       """

        # Debug flag
        self._debugEnabled = debug

        # Last prompt
        self._prompt = ''

        # File streams to/from the emulator
        self._to3270 = None
        self._from3270 = None

    def __del__(self):
        self._Debug('_x3270if deleted')

    @property
    def prompt(self):
        """Gets the last emulator prompt
           str: Last emulator prompt

        """
        return self._prompt

    def Run(self,cmd,*args):
        """Send an action to the emulator

           Args:
              cmd (str): Action name
                 If 'args' is omitted, this is the entire command and the text
                 will be passed through unmodified.
              args (iterable): Arguments
           Returns:
              str: Command output
                 Mulitiple lines are separated by newline characters.
           Raises:
              x3270if.ActionFailException: Action failed.
              EOFException: Emulator exited unexpectedly.
        """
        if (not isinstance(cmd, str)):
            raise Exception("First argument must be a string")
        self._Debug("args is {0}, len is {1}".format(args, len(args)))
        if (args == ()):
            argstr = cmd
        elif (len(args) == 1 and not isinstance(args[0], str)):
            # One argument that can be iterated over.
            argstr = cmd + '(' + ','.join(Quote(arg) for arg in args[0]) + ')'
        else:
            # Multiple arguments.
            argstr = cmd + '(' + ','.join(Quote(arg) for arg in args) + ')'
        self._to3270.write(argstr + '\n')
        self._to3270.flush()
        self._Debug('Sent ' + argstr)
        result = ''
        prev = ''
        while (True):
            text = self._from3270.readline().rstrip('\n')
            if (text == ''): raise EOFError('Emulator exited')
            self._Debug("Got '" + text + "'")
            if (text == 'ok'):
                self._prompt = prev
                break
            if (text == 'error'):
                self._prompt = prev
                raise ActionFailException(result)
            if (result == ''): result = prev.lstrip('data: ')
            else: result = result + '\n' + prev.lstrip('data: ')
            prev = text
        return result

    def _Debug(self,text):
        """Debug output

           Args:
              text (str): Text to log. A Newline will be added.
        """
        if (self._debugEnabled):
            if os.name != 'nt':
                sys.stderr.write('[33m')
            sys.stderr.write(text + '[0m\n')
            if os.name != 'nt':
                sys.stderr.write('[0m')
            sys.stderr.write('\n')

class WorkerConnection(_x3270if):
    """Connection to the emulator from a worker script invoked via the Script() action"""
    def __init__(self,debug=False):
        """Initialize the object.

           Args:
              debug (bool): True to log debug information to stderr.

           Raises:
              StartupException: Insufficient information in the environment to
              connect to the emulator.
        """
        _x3270if.__init__(self, debug)
        self._socket = None
        self._infd = -1
        self._outfd = -1

        # Socket or pipes
        port = os.getenv('X3270PORT')
        if (port != None):
            # Connect to a TCP port.
            self._socket = socket.create_connection(['127.0.0.1',int(port)])
            self._to3270 = self._socket.makefile('w', encoding='utf-8')
            self._from3270 = self._socket.makefile('r', encoding='utf-8')
            self._Debug('Connected')
            emulatorEncoding = self.Run('Query(LocalEncoding)')
            if (emulatorEncoding != 'UTF-8'):
                self._to3270 = self._socket.makefile('w',
                        encoding=emulatorEncoding)
                self._from3270 = self._socket.makefile('r',
                        encoding=emulatorEncoding)
        else:
            # Talk to pipe file descriptors.
            infd = os.getenv('X3270INPUT')
            outfd = os.getenv('X3270OUTPUT')
            if (infd == None or outfd == None):
                raise StartupException("No X3270PORT, X3270INPUT or X3270OUTPUT defined")
            self._infd = int(infd)
            self._to3270 = io.open(self._infd, 'wt', encoding='utf-8',
                    closefd=False)
            self._outfd = int(outfd)
            self._from3270 = io.open(self._outfd, 'rt', encoding='utf-8',
                    closefd=False)
            self._Debug('Pipes connected')
            emulatorEncoding = self.Run('Query(LocalEncoding)')
            if (emulatorEncoding != 'UTF-8'):
                self._to3270 = io.open(self._infd, 'wt',
                        encoding=emulatorEncoding, closefd=False)
                self._from3270 = io.open(self._outfd, 'rt',
                        encoding=emulatorEncoding, closefd=False)

    def __del__(self):
        if (self._socket != None): self._socket.close();
        if (self._infd != -1): os.close(self._infd)
        if (self._outfd != -1): os.close(self._outfd)
        _x3270if.__del__(self)
        self._Debug('WorkerConnection deleted')

_badHostChars = '@,[]='
_goodLuChars = 'ABCDEFGHIJKLMNOPQRSTUVWYZabcdefghijklmnopqrstuvwxyz0123456789_-'
_badAcceptChars = '@,[]=:'

class NewEmulator(_x3270if):
    """Starts a new copy of s3270"""
    def __init__(self,debug=False,extra_args=[]):
        """Initialize the object.

           Args:
              debug (bool): True to log debug information to stderr.
              extra_args(list of str, optional): Extra arguments
                 to pass in the s3270 command line.
           Raises:
              StartupException: Unable to start s3270.
        """
        _x3270if.__init__(self, debug)
        self._socket = None
        self._s3270 = None

        # Create a temporary socket to find a unique local port.
        tempsocket = socket.socket()
        tempsocket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        tempsocket.bind(('127.0.0.1', 0))
        port = tempsocket.getsockname()[1]
        self._Debug('Port is {0}'.format(port))

        # Create the child process.
        try:
            args = ['s3270' if os.name != 'nt' else 'ws3270.exe',
                    '-utf8',
                    '-minversion', '3.6',
                    '-scriptport', str(port),
                    '-scriptportonce'] + extra_args
            oserr = None
            try:
                self._s3270 = subprocess.Popen(args,
                        stderr=subprocess.PIPE,universal_newlines=True)
            except OSError as err:
                oserr = str(err)

            if (oserr != None): raise StartupException(oserr)

            # It might take a couple of tries to connect, as it takes time to
            # start the process. We wait a maximum of half a second.
            tries = 0
            connected = False
            while (tries < 5):
                try:
                    self._socket = socket.create_connection(['127.0.0.1', port])
                    connected = True
                    break
                except:
                    time.sleep(0.1)
                    tries += 1
            if (not connected):
                errmsg = 'Could not connect to emulator'
                self._s3270.terminate()
                r = self._s3270.stderr.readline().rstrip('\r\n')
                if (r != ''): errmsg += ': ' + r
                raise StartupException(errmsg)

            self._to3270 = self._socket.makefile('w', encoding='utf-8')
            self._from3270 = self._socket.makefile('r', encoding='utf-8')
            self._Debug('Connected')
        finally:
            del tempsocket

    def __del__(self):
        if (self._s3270 != None): self._s3270.terminate()
        if (self._socket != None): self._socket.close();
        _x3270if.__del__(self)
        self._Debug('NewEmulator deleted')

class HostSpecification:
    """Host specification with proper formatting"""
    def __init__(self,hostName,Port=23,LogicalUnitNames=[],TlsTunnel=False,ValidateHostCertificate=True,AcceptName=None):
        """Initialize an instance

           Args:
              hostName (str): Host name or IP address.
              Port (int, optional): TCP port number.
              LogicalUnitNames (list of str, optional): Logical unit names.
              TlsTunnel (bool, optional): Set up a TLS tunnel.
              ValidateHostCertificate (bool, optional): Validate the host TLS certificate.
              AcceptName (str, optional): Host name to accept in the host TLS certificate.
       """
        self.HostName = hostName
        self.Port = Port
        self.LogicalUnitNames = LogicalUnitNames
        self.TlsTunnel = TlsTunnel
        self.ValidateHostCertificate = ValidateHostCertificate
        self.AcceptName = AcceptName

    @property
    def HostName(self):
        """Host name or IP address"""
        return self._hostName
    @HostName.setter
    def HostName(self,value):
        if (any(ch in value for ch in _badHostChars)):
            raise Exception("HostName contains invalid character(s)")
        self._hostName = value

    @property
    def Port(self):
        """TCP port number"""
        return self._port
    @Port.setter
    def Port(self,value):
        self._port = int(value)
        if (self._port < 1 or self._port > 0xffff):
            raise Exception("Invalid port value")

    @property
    def LogicalUnitNames(self):
        """List of Logical Unit (LU) names"""
        return self._logicalUnitNames
    @LogicalUnitNames.setter
    def LogicalUnitNames(self,value):
        for lu in value:
            if (any(ch not in _goodLuChars for ch in lu)):
                raise Exception("Logical unit name contains invalid character(s)")
        self._logicalUnitNames = value

    @property
    def AcceptName(self):
        """Name to accept in the host TLS certificate"""
        return self._acceptName
    @AcceptName.setter
    def AcceptName(self,value):
        if (value == None):
            self._acceptName = value
            return
        for lu in value:
            if (any(ch in value for ch in _badAcceptChars)):
                raise Exception("Accept name contains invalid character(s)")
        self._acceptName = value

    def __str__(self):
        if (self.HostName == None):
            return ''
        r = ''
        if (self.TlsTunnel):
            r += "L:"
        if (not self.ValidateHostCertificate):
            r += "Y:"
        if (self.LogicalUnitNames != []):
            r += ','.join(self.LogicalUnitNames) + '@'
        if (':' in self.HostName):
            r += '[' + self.HostName + ']'
        else:
            r += self.HostName
        if (self.Port != 23):
            r += ':' + str(self.Port)
        if (self.AcceptName != None):
            r += '=' + self.AcceptName
        return r

